웹소켓을 이용한 채팅 메세지 전송 기능 구현

kimcno3 opened the pull request.

kimcno3 committed:
Redis 서버 연결 및 RedisTemplate 객체 주입

* Redis 의존성 추가
* RedisConfig 클래스 생성
* Redis docker 포트 변경
* ChatHandler에 RedisTemplate 객체 주입
* 소켓 연결 및 해지 시 메소드에 redis 메소드 추가

kimcno3 committed:
RestTemplate 적용한 메세지 전송 API 구현

* RestTemplate 빈 객체 생성
* channelId에 관련된 participants를 조회하는 메소드 생성
* 메세지 전송의 대상이 되는 회원 정보 조회 메소드 생성

kimcno3 committed:
메세지 대상 유저 조회 API 수정

* Controller 와 Service 레이어의 책임 구분 재정의

kimcno3 committed:
메세지 대상 유저 조회 API 변경을 위한 코드 수정

* message 전송 API로 목적을 변경
* webSocketSession에서 토큰을 조회해 userId조회 메소드를 구현

f-lab-bright commented on src/main/java/com/flab/football/config/ChatConfig.java:
```diff
@@ -21,9 +21,9 @@ public class ChatConfig {
 
   @Bean
-  public List<WebSocketSession> sessions() {
+  public Map<String, WebSocketSession> sessions() {
```
key를 string으로 사용하는 이유가 있나요?
user_id는 numeric이지 않나요??

kimcno3 replied:
setKeySerializer를 GenericToStringSerializer로 지정하면서 Integer Type의 키를 사용하면서 Redis에 저장되는 타입은 String이 되도록 변경했습니다.

아무런 설정없이 Integer를 키 타입으로 지정하면 Redis에서 조회할 경우에 직렬화된 형태로 조회가 되어 확인하는데 어려움을 느껴 선택한 직렬화 방식입니다!

f-lab-bright commented on src/main/java/com/flab/football/controller/ChatController.java:
```diff
@@ -68,4 +79,36 @@ public ResponseDto findChannels() {
+        log.info(userId + "님에게 푸시 알림을 보냅니다.");
+
+      } else {
+
+        String uri = "ws:/" + localAddress + "/ws/chat";
+
+        // 해당 소켓 서버에 어떻게 메세지를 전달해 줄 수 있을 지 고민해봐야 한다.
+
```
`WebSocketSession` 을 이용해서 연결된 클라이언트에게 메세지를 보내야 합니다 :)

f-lab-bright commented on src/main/java/com/flab/football/handler/ChatHandler.java:
```diff
@@ -55,42 +76,29 @@ protected void handleTextMessage(WebSocketSession session, TextMessage message)
-     */
+    String userId = String.valueOf(securityService.getCurrentUserId(bearerToken));
+
+    // userId 와 웹소켓 서버 정보를 redis 에 저장
+    redisTemplate.opsForValue().set(userId, session.getLocalAddress().toString());
+
+    // 웹소켓 서버 내 메모리에 session 객체를 저장
+    sessions.put(userId, session);
```
이 부분은 추후에 서비스가 분리되면, 직접 redis에 접근하지 않을 수 있기 때문에 추상화 레이어를 하나 감싸면 더 좋겠네요 :)

kimcno3 replied:
RedisSerivce 인터페이스 및 구현 클래스를 선언해 redis에 접근하는 책임을 담당하는 서비스 레이어를 추가했습니다!

f-lab-bright commented on src/main/java/com/flab/football/handler/ChatHandler.java:
```diff
@@ -55,42 +76,29 @@ protected void handleTextMessage(WebSocketSession session, TextMessage message)
-       * sessionList에 존재하는 user라면 접속중인 user 이므로 message를 전송한다.
-       * sessionList에 존재하지 않는다면 접속중이진 않지만 channel에 포함된 user이므로 푸시알림의 대상이 된다.
-     */
+    session.sendMessage(message);
```
요 지점은 client에게서 websocket으로 메세지를 받는 지점이지 않나요..??
여기에서는 메세지 받는 처리를 할 필요는 없을 것 같네요 :)

f-lab-bright reviewed (commented):
서버가 실행중이면, 현재 나의 서버의 주소를 redis에 등록해두는게 필요할 것 같은데, 그 부분이 빠져있는 것 같네요..?
그리고 websocket의 경우에도 커넥션이 끊어졌지만, 서버에서 파악하지 못하고 있는 상태로 남을 수 있는데, 이 부분은 어떻게 해결할 수 있을지도 한 번 고민해보시면 좋을 것 같습니다 :)

kimcno3 committed:
피드백 반영

* Redis 접근 메소드에 대한 책임을 가질 추상화 레이어 추가
  - RedisService, RedisServiceImpl 클래스 생성
  - DI 객체 변경
* RedisTemplate key Type을 Integer로 변경
  - keySerializer를 GenericToStringSerializer 클래스로 지정
* Message 엔티티 클래스 및 테이블에 보낸 User에 대한 정보를 추가하기 위해 외래키 생성

* RestTemplate을 활용해 ChatHandler 의 handleTextMessage() 호출을 담당하는 API 구현

kimcno3 commented:
컨트롤러 레이어에서 웹소켓 서버로 메세지 정보를 어떻게 전달해줘야 하는건지에 대한 해답을 찾지 못한 상태라 일부 피드백만 적용해 봤습니다ㅠㅠ

구현을 하다보니 의도와는 다르게 작성되는 느낌이 강한 것 같습니다...어디서부터 잘못 접근한건지도 모르겠어서 어떤 질문을 드려야할 지도 생각하지 못하는 것 같아 의도했던 설계의 동작 흐름을 다시 집어보면서 잘못 구현된 부분을 찾아보려고 합니다...ㅠㅠ

특히나 서버를 분리한다는 개념 자체가 코드 내에서 어떻게 구분해줘야 하는지를 찾지 못해서 잘못된 방향으로 코드를 작성하는 건 아닌지 생각이 듭니다ㅠㅠ

kimcno3 committed:
리팩토링

* WebSocket 패키지를 다른 패키지와 분리
  - 이후 모듈로 분리할 대상을 패키지로 미리 분류한 개념입니다.

* restTemplate 에 의해 호출되는 API에 대한 인증 권한을 허가
  - 추후 보안해야 되는 사항입니다.

kimcno3 commented:
websocket 패키지를 생성해 소켓에서 발생하는 채팅 기능에 대한 책임을 다른 패키지오 분리시켰습니다!

제 생각에 이후에 해야 할 작업은 **실제로 사용자로부터 받은 메세지 정보를 어떻게 매핑해 RestTemplate으로 해당 소켓 서버로 전송해줄 지** 고민해봐야할 것 같습니다! 현재 코드는 테스트용으로 "Hello" 라는 메세지를 직접 TextMessage에 매개변수로 넘겨 전송하고 있는 상태입니다.

테스트하면서 사용한 포트는 다음과 같습니다.

- API 서버는 8080 포트로 고정했고 회원 로그인 요청과 메세지 전송 요청을 처리하도록 했습니다.
- WebSocket 서버는 8081과 8082 포트로 두 서버를 사용했고 로그인된 회원의 토큰으로 인가를 허가받고 소켓에 연결했습니다.

kimcno3 committed:
RestTemplate 호출 API 유형 변경

* post -> get

kimcno3 committed:
리팩토링

* ChatHandler.handleTextMessage() 구현
  - 웹소켓에 연결된 사용자가 메세지 전송을 요청
  - API 서버에 수신 대상 사용자 조회 요청

* SessionConfig
  - ObjectMapper 빈 객체 생성 메소드 추가

* SecurityConfig
  - API 인가 허용 처리 추가

* SendMessageOrPushRequest, SendMessageRequest
  - Json Request 매핑 객체 추가

* GET -> POST 로 변경
*

kimcno3 committed:
리팩토링

* ChatHandler에서 메세지 전송 메소드 삭제
* Message 엔티티 객체 DB 저장 로직 추가
* Message 엔티티 객체 빌드시 Type 값 지정
* 필드명 변경

kimcno3 committed:
피드백 반영

* API 서버와 WebSocket 서버의 명확한 책임 분리를 위한 리팩토링
  - PathVariable 삭제 : Request 객체를 통한 데이터 전송
  - ChatPushService 구현 : 로직의 단위를 세분화
  - Request, command 객체 추가 생성
  - sendUser는 push 하지 않도록 조건문 추가

f-lab-bright commented on src/main/java/com/flab/football/config/SecurityConfig.java:
```diff
@@ -57,6 +57,7 @@ protected void configure(HttpSecurity http) throws Exception {
           .and()
           .authorizeRequests()
           .antMatchers("/user/signup", "/user/login").permitAll()
+          .antMatchers("/ws/send/message").permitAll() // 인증 방법을 찾아야 한다.
```
hint를 몇가지 드리자면...
1. WebSecurityConfigurerAdapter 를 여러개를 상속받아서 설정할 수 있습니다.
2. void configure(HttpSecurity http) 메소드에서 requestMatcher 를 이용하면 여기에 매칭되는 uri에 대해서만 설정을 적용할 수 있습니다.
3. 심플하게 인증을 적용해보신다면 basic auth 방식을 써보셔도 좋겠네요 :)

kimcno3 replied:
제가 선택한 방법은 WebSecurity 객체에 대한 설정을 할 수 있는 configure() 메소드를 오버라이딩해 내부 호출 API에 대해서 인증, 인가처리에 대해 제외하도록 수정해봤습니다!

webSecurity에 대한 설정을 할 경우, 해당 엔드포인트에 대해선 Spring Security의 filter 대상에서 제외시킨다고 알고 있습니다!

내부 호출 API의 경우 인증, 인가에 대한 필요성이 없을 가능성이 높다고 판단했고, 그렇다면 이에 대한 절차에서 처음부터 제외시켜 주는 것이 불필요한 로직 수행을 방지할 수 있을 것이라 생각해 위 방법으로 구현해봤습니다!

f-lab-bright commented on src/main/java/com/flab/football/controller/ChatController.java:
```diff
@@ -68,4 +76,21 @@ public ResponseDto findChannels() {
+  @PostMapping("/send/message")
+  public ResponseDto sendMessage(@RequestBody SendMessageRequest request) {
+
+    // 메세지를 보내는 사용자 id 조회
+    int sendUserId = securityService.getCurrentUserId();
```
@AuthenticationPrincipal 을 이용할 수 있으면 더 좋겠네요.

kimcno3 replied:
@AuthenticationPrincipal 에 대한 구현은 다른 브랜치로 구분해서 구현해봐도 괜찮을지 궁금합니다!!

f-lab-bright replied:
그럼요 :)

f-lab-bright commented on src/main/java/com/flab/football/service/chat/ChatPushServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+    } else {
+
+      // 그 외 경우엔 해당 서버에 접속중인 회원이 Map 컬렉션에 저장되어 있기에 메세지를 전송한다.
+      String uri = "http:/" + address + "/ws/send/message";
```
address가 /localhost:8081 처럼 들어가있나요..?
아니면 http:/ 가 아니라 http:// 가 되어야 할 것 같습니다 :)

kimcno3 replied:
네! `/127.0.0.1:8081` 형식으로 들어갑니다! 해당 uri로 정상적으로 API가 호출됩니다😁

f-lab-bright commented on src/main/java/com/flab/football/service/chat/ChatServiceImpl.java:
```diff
@@ -78,25 +80,46 @@ public void inviteParticipants(int channelId, List<Integer> participants) {
+
+        command.setReceiveUserId(receiveUserId);
+
+        chatPushService.pushMessage(command);
```
이 코드 안전할까요??
위에서 command을 하나만 만들어두고, receiveUserId를 바꿔가면서 pushMessage에 전달해주는데, pushMessage는 또 비동기 메소드라 동시에 둘 이상의 처리가 될 수도 있겠네요.
그럼 receiveUserId가 모든 pushMessage 메소드 내에서 사용될 때 요청에 전달 된 receiveUserId가 변경되지 않고 사용될걸 보장할 수 있을까요??

kimcno3 replied:
비동기 처리에 대한 고려를 하지 않고 동일한 값에 대한 재활용에 대한 고민만 했던 것 같습니다ㅠㅠ

비동기 처리 시 동시성에 대한 고민을 해본다면 일단 비동기 처리 메소드가 호출되기 전에 해당 receiveUserId에 대한 command 객체를 생성하도록 수정했습니다!

f-lab-bright commented on src/main/java/com/flab/football/websocket/handler/ChatHandler.java:
```diff
@@ -0,0 +1,78 @@
+    int userId = getCurrentUserId(session);
+
+    // userId 와 웹소켓 서버 정보를 redis 에 저장
+    redisService.setValue(userId, session.getLocalAddress().toString());
```
userId만 key로 사용하는건 위험할 수도 있겠네요.
보통은 키에 prefix같은걸 같이 이용하면 좋습니다.
user::123 or user::v1::123 같은 형태로 사용하게 되면, 새로운 키셋으로 변경할 때도 유리하고, 버저닝같은걸 사용하기도 편해집니다 :)

kimcno3 replied:
수정했습니다! 

prefix 값은 utils 클래스를 생성해 선언해 websocket 패키지 내에서 활용하도록 리펙토링했습니다!

f-lab-bright commented on src/main/java/com/flab/football/websocket/handler/ChatHandler.java:
```diff
@@ -0,0 +1,78 @@
+    int userId = getCurrentUserId(session);
+
+    // userId 와 웹소켓 서버 정보를 redis 에 저장
+    redisService.setValue(userId, session.getLocalAddress().toString());
```
주소를 session에서 getLocalAddress에서 가져오는 것 보다, 해당 서버가 실행 될 때 주소를 미리 읽어와두고, 계속 재활용하는게 좋지 않을까요?

kimcno3 replied:
빈 객체 등록 과정에서 서버 주소와 포트 번호를 조회해 하나의 빈 객체로 등록해 재활용하도록 수정해봤습니다!

kimcno3 committed:
내부 API 인가 무시 처리 코드 작성

* SecurityConfig.configure() 오버라이딩
  - webSecurity 관련 설정 추가

kimcno3 committed:
비동기 메소드 처리 방식을 고려한 command 객체 생성 시점 변경

* 비동기 메소드가 호출될 때마다 하나의 command 객체를 생성해 전달
* 객체 공유로 인한 데이터 불일치 문제 해결
* 불필요한 setter 메소드 삭제

kimcno3 committed:
Redis prefix key 추가 및 RedisService 메소드명 변경

* key에 prefix 추가
  - prefix 값을 저장 및 활용하기 위한 utils 클래스 생성

* Redis key 타입 변경
  - prefix 추가로 인한 타입 변경
  - Integer -> String

* RedisService 메소드명 변경
  - 기존 메소드명보다 직관적이게 변경

kimcno3 committed:
서버 address 멤버 변수 선언

* 서버 실행시 서버 주소 정보를 조회해 멤버 변수로 저장
* 저장된 address 값을 재활용
* ChatPushServiceImpl -> uri 문자열 수정

kimcno3 committed:
서버 주소 저장 객체를 빈 객체로 등록

kimcno3 committed:
WebSocketSession을 다루는 Service 레이어 추가
